src/engine/loopingcontrol.cpp

```diff
@@ -711,151 +820,280 @@ void LoopingControl::clearActiveBeatLoop() {
     }
 }
 
-void LoopingControl::slotBeatLoop(double beats, bool keepStartPoint) {
-    int samples = m_pTrackSamples->get();
-    if (!m_pTrack || samples == 0) {
-        clearActiveBeatLoop();
-        return;
+bool LoopingControl::currentLoopMatchesBeatloopSize() {
+    LoopSamples loopSamples = m_loopSamples.getValue();
+    int currentLoopLength = loopSamples.end - loopSamples.start;
+
+    double beatloopSize = m_pCOBeatLoopSize->get();
+    int beatLoopLength = calculateEndOfBeatloop(loopSamples.start, beatloopSize)
+                         - loopSamples.start;
+
+    // qDebug() << "LoopingControl::currentLoopMatchesBeatloopSize" << currentLoopLength << beatLoopLength;
+
+    // calculateEndOfBeatloop will round to the nearest even sample, so allow
+    // 2 samples of wiggle room.
+    return abs(currentLoopLength - beatLoopLength) <= 2;
+}
+
+int LoopingControl::calculateEndOfBeatloop(int startSample, double beatloopSize) {
+    if (m_pBeats == nullptr) {
+        return kNoTrigger;
     }
 
-    if (!m_pBeats) {
-        clearActiveBeatLoop();
-        return;
+    int fullbeats = static_cast<int>(beatloopSize);
```

Comment:
I'm not entirely sure. This is how the code was previously when it was [in the slotBeatloop function](https://github.com/mixxxdj/mixxx/pull/1187/files/0ae880567fc462b58fb68bcd9269bc1c6c4a9445#diff-13add587f80110bacb905e9f6cf0eea7L795).